File: src/implementations/FieldOfView.cpp
//
// Created by wint3rmute on 3/28/18.
//

#include <iostream>
#include "../headers/FieldOfView.h"


FieldOfView::FieldOfView(Plane *plane, int howManyCells) : GameObject(0)  {

    this->plane = plane;
    visualCellsCount = howManyCells;
    objectType = HIDEO_KOJIMA;

    cells = new sf::CircleShape * [visualCellsCount];

    for (int i = 0; i < visualCellsCount; ++i) {
        cells[i] = new sf::CircleShape(VISUAL_CELL_SIZE);

        cells[i]->setOrigin(2,2);
    }

    distances_to_bullets = new double [howManyCells];
    distances_to_planes = new double [howManyCells];


}

void FieldOfView::simulate() {

    if(plane->wasRemoved) //
    {
        GameEngine::removeObject(this);
        return;
    }


    for (int i = 0; i < visualCellsCount; ++i) {
        cells[i]->setPosition(plane->getPosition());


        cells[i]->move(
                (cos((360.0f / visualCellsCount * i + plane->getRotation()) * DEG_2_RAD)* VISUAL_CELL_DISTANCE_FROM_PLANE),
                (sin((360.0f / visualCellsCount * i + plane->getRotation()) * DEG_2_RAD)* VISUAL_CELL_DISTANCE_FROM_PLANE)
        );
    }


    double smallestDistanceBullet = 100000.0,
            smallestDistancePlane = 100000.0;

    int bestVisualCellNumberBullet = -1,
            bestVisualCellNumberPlane = -1;

    for (int j = 0; j < visualCellsCount; ++j) {
        distances_to_bullets[j] = 0.0;
        distances_to_planes[j] = 0.0;
    }


    for(auto gameObject : GameEngine::gameObjects) {

        if(gameObject->collisionMode == NON_COLLIDING or gameObject == this->plane)
            continue;

        for (int i = 0; i < visualCellsCount; ++i) {

            double temporaryDistance = GameEngine::getDistance(gameObject->getPosition(), cells[i]->getPosition());

            if(gameObject->collisionMode == AFFECTOR) { //if we have a bullet TODO: this if fires too many times, this could be checked before the loop
                if (temporaryDistance < smallestDistanceBullet) {
                    smallestDistanceBullet = temporaryDistance;
                    bestVisualCellNumberBullet = i;
                }

            } else {

                if (temporaryDistance < smallestDistancePlane) { //if we have a plane
                    smallestDistancePlane = temporaryDistance;
                    bestVisualCellNumberPlane = i;
                }
            }
        }

        if(bestVisualCellNumberPlane != -1)
            distances_to_planes[bestVisualCellNumberPlane] = smallestDistancePlane;


        if(bestVisualCellNumberBullet != -1)
            distances_to_bullets[bestVisualCellNumberBullet] = smallestDistanceBullet;
    }

    for (int i = 0; i < visualCellsCount; ++i) {
        cells[i]->setFillColor(sf::Color( 56 +distances_to_planes[i] * 100 , 56 + distances_to_bullets[i] * 100, 56));
        //std::cout << distances_to_planes[i] << std::endl;
    }




}

void FieldOfView::draw(sf::RenderTarget &target, sf::RenderStates states) const {

    for (int i = 0; i < visualCellsCount; ++i) {
        target.draw(*cells[i]);
    }
}

sf::Vector2f FieldOfView::getPosition() {
    return plane->getPosition();
}

double FieldOfView::getRotation() {
    return plane->getRotation();
}

FieldOfView::~FieldOfView() {


    for (int i = 0; i < visualCellsCount; ++i) {
        delete cells[i];

        delete cells;

        delete distances_to_bullets;
        return;

        delete distances_to_planes;
}

}

File: python-planes/FieldOfView.py
import GameObject
import constants
import pymunk
import math
import Plane
import Bullet


def get_distance(position_a, position_b):
    return math.sqrt((position_a[0] - position_b[0])**2 + (position_a[1] - position_b[1])**2)


class FieldOfView(GameObject.GameObject):
    def __init__(self, plane_me):
        self.plane_me = plane_me
        self.cells = [None] * constants.VISUAL_CELLS_COUNT

        self.distances_to_bullets = [0] * constants.VISUAL_CELLS_COUNT
        self.distances_to_planes = [0] * constants.VISUAL_CELLS_COUNT

    def remove(self):
        pass

    def update(self):

        for i in range(constants.VISUAL_CELLS_COUNT):
            self.cells[i] = [
                self.plane_me.body.position.x + math.cos((360.0 / constants.VISUAL_CELLS_COUNT * i + self.plane_me.body.angle * constants.RAD_2_DEG) * constants.DEG_2_RAD) * constants.VISUAL_CELL_DISTANCE_FROM_PLANE,
                self.plane_me.body.position.x + math.sin((360.0 / constants.VISUAL_CELLS_COUNT * i + self.plane_me.body.angle * constants.RAD_2_DEG) * constants.DEG_2_RAD) * constants.VISUAL_CELL_DISTANCE_FROM_PLANE
            ]

            self.distances_to_bullets[i] = 0.
            self.distances_to_planes[i] = 0.

        smallest_distance_bullet, smallest_distance_plane = 10000, 10000
        best_visual_cell_number_bullet, best_visual_cell_number_plane = -1, -1

        for game_object in self.game_engine.game_objects:
            if game_object == self or game_object == self.plane_me:
                continue

            for i in range(len(self.cells)):
                temp_distance = get_distance(game_object.body.position, self.cells[i])
                if type(game_object) is Plane.Plane:
                    if temp_distance < smallest_distance_plane:
                        smallest_distance_plane = temp_distance
                        best_visual_cell_number_plane = i

                elif type(game_object) is Bullet.Bullet:
                    if temp_distance < smallest_distance_bullet:
                        smallest_distance_bullet = temp_distance
                        best_visual_cell_number_bullet = i

            if best_visual_cell_number_bullet != -1:
                self.distances_to_bullets[best_visual_cell_number_bullet] = smallest_distance_bullet

            if best_visual_cell_number_plane != -1:
                self.distances_to_planes[best_visual_cell_number_plane] = smallest_distance_plane
